README.md
# sphinxsearch

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

**Note:** Replace ```Gordon Anderson``` ```gordonbanderson``` ```https://github.com/gordonbanderson``` ```gordon.b.anderson@gmail.com``` ```suilven``` ```sphinxsearch``` ```Search SilverStripe using Sphinx``` with their correct values in [README.md](README.md), [CHANGELOG.md](CHANGELOG.md), [CONTRIBUTING.md](CONTRIBUTING.md), [LICENSE.md](LICENSE.md) and [composer.json](composer.json) files, then delete this line. You can run `$ php prefill.php` in the command line to make all replacements at once. Delete the file prefill.php as well.

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what
PSRs you support to avoid any confusion with users and contributors.

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practices by being named the following.

```
bin/        
config/
src/
tests/
vendor/
```


## Install

Via Composer

``` bash
$ composer require suilven/sphinxsearch
```

## Usage

``` php
$skeleton = new Suilven\SphinxSearch();
echo $skeleton->echoPhrase('Hello, League!');
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email gordon.b.anderson@gmail.com instead of using the issue tracker.

## Credits

- [Gordon Anderson][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/suilven/sphinxsearch.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/suilven/sphinxsearch/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/suilven/sphinxsearch.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/suilven/sphinxsearch.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/suilven/sphinxsearch.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/suilven/sphinxsearch
[link-travis]: https://travis-ci.org/suilven/sphinxsearch
[link-scrutinizer]: https://scrutinizer-ci.com/g/suilven/sphinxsearch/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/suilven/sphinxsearch
[link-downloads]: https://packagist.org/packages/suilven/sphinxsearch
[link-author]: https://github.com/gordonbanderson
[link-contributors]: ../../contributors


CHANGELOG.md
# Change Log

## [0.0.3](https://github.com/gordonbanderson/sphinxsearch/tree/0.0.3) (2018-04-03)
[Full Changelog](https://github.com/gordonbanderson/sphinxsearch/compare/0.0.2...0.0.3)

**Closed issues:**

- Remove Debug [\#3](https://github.com/gordonbanderson/sphinxsearch/issues/3)

## [0.0.2](https://github.com/gordonbanderson/sphinxsearch/tree/0.0.2) (2018-04-03)
[Full Changelog](https://github.com/gordonbanderson/sphinxsearch/compare/0.0.1...0.0.2)

## [0.0.1](https://github.com/gordonbanderson/sphinxsearch/tree/0.0.1) (2018-03-24)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tests/ClientTest.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 24/3/2561
 * Time: 20:36 น.
 */

namespace Suilven\SphinxSearch\Tests;


use SilverStripe\Dev\SapphireTest;
use Suilven\FreeTextSearch\Indexes;
use Suilven\SphinxSearch\Service\Client;
use Suilven\SphinxSearch\Service\Indexer;

class ClientTest extends SapphireTest
{
    public function testClient()
    {
        $indexesService = new Indexes();
        $indexes = $indexesService->getIndexes();
        $indexer = new Indexer($indexes);
        $indexer->saveConfig();
    }
}


tests/SphinxTest.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 24/3/2561
 * Time: 0:51 น.
 */

namespace Suilven\SphinxSearch\Tests;


use Foolz\SphinxQL\Facet;
use Foolz\SphinxQL\Helper;
use Model\Photo;
use SilverStripe\Control\Director;
use SilverStripe\Core\Config\Config;
use SilverStripe\Dev\SapphireTest;

use Foolz\SphinxQL\SphinxQL;
use Foolz\SphinxQL\Connection;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\DB;
use Suilven\FreeTextSearch\Indexes;
use Suilven\SphinxSearch\Service\Client;
use Suilven\SphinxSearch\Service\Indexer;
use Suilven\SphinxSearch\Service\Searcher;

class SphinxTest extends SapphireTest
{
    protected static $fixture_file = 'sphinxsearch/fixtures.yml';

    const INDEX_NAME = 'photographs';


    protected static $extra_dataobjects = [
        Photo::class
    ];


    /**
     * Recreate the sphinx index clean each time
     */
    public function setUp()
    {
        parent::setUp();

        $indexes = [
            [
                // @todo, this does not look right
                'index' => [
                    'name' => self::INDEX_NAME,
                    'class' => 'Model\Photo',
                    'fields' => [
                        'Title',
                        'Description'
                    ]
                ]
            ]
        ];

        Config::nest();

        Config::inst()->update('Suilven\FreeTextSearch\Indexes', 'indexes', $indexes);

        $cf = Config::inst()->get('Suilven\FreeTextSearch\Indexes', 'indexes');

        $all = Config::inst()->getAll();
        error_log('CF: ' . print_r($cf, 1));

        // seems to be using the default

        error_log('IS DEV? ' . Director::isDev());
        error_log('IS TEST? ' . Director::isTest());


        // override index definitions for testing
       // Config::inst()->nest();



        $database = DB::get_conn()->getSelectedDatabase();
        $databaseHost = DB::get_conn()->getDatabaseServer();

        error_log('TEMP DB: ' . print_r($database, 1));
        error_log('TEMP DB HOST: ' . print_r($databaseHost, 1));

        // save config
        $indexesService = new Indexes();
        $indexesObj = $indexesService->getIndexes();
        $indexer = new Indexer($indexesObj);
        $indexer->setDatabaseName($database);
        $indexer->saveConfig();

        // perhaps should be Server instead of Client
        $client = new Client();
        $client->reindex();

    }

    public function test_search()
    {

        foreach(Photo::get() as $photo) {
            error_log('FROM DB: ' . $photo->Title);
        }


        $searcher = new Searcher();
        $searcher->setIndex(self::INDEX_NAME);
        $results = $searcher->search('Central Bangkok');
        error_log(print_r($results, 1));

    }
}


tests/Model/Photo.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 9/4/2561
 * Time: 11:26 น.
 */

namespace Model;


use SilverStripe\Dev\TestOnly;
use SilverStripe\ORM\DataObject;

class Photo extends DataObject implements TestOnly
{
    private static $db = array(
        'Title' => 'Varchar(255)',
        'Description' => 'HTMLText',
        'TakenAt' => 'Datetime',

        'GeoIsPublic' => 'Boolean',

        // these should come from mappable but it's not 4.x ready
        'Lat' => 'Decimal(18,15)',
        'Lon' => 'Decimal(18,15)',

        'FlickrPlaceID' => 'Varchar(255)',

        'Aperture' => 'Float',
        'ShutterSpeed' => 'Varchar',
        'FocalLength35mm' => 'Int',
        'ISO' => 'Int',
    );
}


src/Service/Client.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 24/3/2561
 * Time: 19:58 น.
 */

namespace Suilven\SphinxSearch\Service;


use SilverStripe\Core\Config\Config;

class Client
{
    /**
     * Get a connection to sphinx using the values configured in YML files for port and host
     *
     * @return Foolz\SphinxQL\Drivers\Pdo\Connection connection object to Sphinx
     */
    public function getConnection()
    {
        $host = Config::inst()->get('Suilven\SphinxSearch\Service\Client', 'host');
        $port = Config::inst()->get('Suilven\SphinxSearch\Service\Client', 'port');

        error_log('HOST: ' . $host);
        error_log('PORT: ' . $port);


        $conn = new \Foolz\SphinxQL\Drivers\Pdo\Connection();
        $conn->setParams(array('host' => $host, 'port' => $port));
        return $conn;
    }

    // see https://stackoverflow.com/questions/6275042/how-to-escape-special-characters-in-sphinxql-fulltext-search
    function escapeSphinxQL ( $string )
    {
        $from = array ( '\\', '(',')','|','-','!','@','~','"','&', '/', '^', '$', '=', "'", "\x00", "\n", "\r", "\x1a" );
        $to   = array ( '\\\\', '\\\(','\\\)','\\\|','\\\-','\\\!','\\\@','\\\~','\\\"', '\\\&', '\\\/', '\\\^', '\\\$', '\\\=', "\\'", "\\x00", "\\n", "\\r", "\\x1a" );
        return str_replace ( $from, $to, $string );
    }

    public function restartServer()
    {
        $restartCommand = Config::inst()->get('Suilven\SphinxSearch\Service\Client', 'cmd_restart');
        exec($restartCommand);
    }


    /**
     * Execute reindex command.  @todo Can this be done using SphinxQL?
     */
    public function reindex()
    {
        $reindexCommand = Config::inst()->get('Suilven\SphinxSearch\Service\Client', 'cmd_reindex');

        // @todo Fix, config not working here
        //$reindexCommand = '/usr/bin/indexer --rotate --all';

        error_log('reindex command ' . $reindexCommand);
        error_log(exec($reindexCommand));
    }
}


src/Service/Searcher.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 25/3/2561
 * Time: 1:35 น.
 */

namespace Suilven\SphinxSearch\Service;


use Foolz\SphinxQL\Drivers\Pdo\ResultSet;
use Foolz\SphinxQL\Facet;
use Foolz\SphinxQL\Helper;
use Foolz\SphinxQL\SphinxQL;
use SilverStripe\ORM\ArrayList;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\PaginatedList;
use SilverStripe\View\ArrayData;
use Suilven\FreeTextSearch\Indexes;

class Searcher
{

    private $client;

    private $pageSize = 10;

    private $page = 1;

    private $index = 'sitetree';

    /**
     * @var array tokens that are facetted, e.g. Aperture, BlogID
     */
    private $facettedTokens = [];

    /**
     * @var array associative array of filters against tokens
     */
    private $filters = [];

    /**
     * @param array $filters
     */
    public function setFilters($filters)
    {
        $this->filters = $filters;
    }

    /**
     * @param int $pageSize
     */
    public function setPageSize($pageSize)
    {
        $this->pageSize = $pageSize;
    }

    /**
     * @param string $index
     */
    public function setIndex($index)
    {
        $this->index = $index;
    }


    /**
     * @param array $facettedTokens
     */
    public function setFacettedTokens($facettedTokens)
    {
        $this->facettedTokens = $facettedTokens;
    }


    /**
     * @param int $page
     */
    public function setPage($page)
    {
        $this->page = $page;
    }

    public function __construct()
    {
        $this->client = new Client();
    }


    public function search($q)
    {
        error_log('SIZE: ' . $this->pageSize);
        error_log('PAGE: ' . $this->page);

        $startMs = round(microtime(true) * 1000);
        $connection = $this->client->getConnection();

        // @todo make fields configurable?
        $query = SphinxQL::create($connection)->select('id')
            ->from([$this->index .'_index', $this->index  . '_rt']);

        if (!empty($q)) {
            $query->match('?', $q);
        }


        // string int fixes needed here
        foreach($this->filters as $key => $value) {
            if ($key !== 'q') {
                if (ctype_digit($value)) {
                    if (is_int($value + 0)) {
                        $value = (int) $value;
                    }
                    else if (is_float($value + 0)) {
                        $value = (float) $value;
                    }
                }

                $query->where($key, $value);
            }
        }

        foreach($this->facettedTokens as $tokenToFacet) {
            $query->facet(Facet::create($connection)
                ->facet(array($tokenToFacet))
                // @todo->orderBy('iso', 'ASC')
            );
        }


        $query->limit(($this->page-1) * $this->pageSize, $this->pageSize);


        /** @var array $result */
        $result = $query->executeBatch()
            ->getStored();

        /** @var ResultSet $resultSet */
        $resultSet = $result[0];

        $ids = $resultSet->fetchAllAssoc();

        $ctr = 1;
        $facets = [];
        foreach($this->facettedTokens as $token)
        {
            $resultSet = $result[$ctr];
            $rawFacets = $resultSet->fetchAllAssoc();
            $tokenFacets = [];
            foreach($rawFacets as $singleFacet) {
                if (isset($singleFacet[$token])) {
                    $value = $singleFacet[$token];
                    $count = $singleFacet['count(*)'];

                    // do this way to maintain order from Sphinx
                    $nextFacet = ['Value' => $value, 'Count' => $count, 'Name' => $token, 'ExtraParam' => "$token=$value"];
                    $filterForFacet = $this->filters;

                    if (isset($this->filters[$token])) {
                        $nextFacet['Selected'] = true;
                        unset($filterForFacet[$token]);
                    } else {
                        // additional value to the URL, unselected facet
                        $filterForFacet[$token] = $value;
                    }

                    // @todo - escaping?
                    $urlParams = '';
                    foreach($filterForFacet as $n => $v) {
                        //if (!isset($this->filters[$token])) {
                            $urlParams .= "{$n}={$v}&";
                       // }
                    }

                    $nextFacet['Params'] = substr($urlParams, 0, -1);

                    $tokenFacets[] = $nextFacet;
                }
            }
            $ctr++;

            // @todo huyman readable title
            $facets[] = ['Name' => $token, 'Facets' => new ArrayList($tokenFacets)];
        }

        $metaQuery = SphinxQL::create($connection)->query('SHOW META;');
        $metaData = $metaQuery->execute();

        $searchInfo = [];
        foreach($metaData->getStored() as $info) {
            $varname = $info['Variable_name'];
            $value = $info['Value'];
            $searchInfo[$varname] = $value;
        }

        $formattedResults = new ArrayList();

        foreach($ids as $assoc) {
            // @todo use array merge to minimize db queries
            // @todo need to get this from the index definition

            $indexesService = new Indexes();
            $indexes = $indexesService->getIndexes();

            $clazz = '';

            // @todo fix this, return an associative array from the above
            foreach($indexes as $indexObj)
            {
                $name = $indexObj->getName();
                if ($name == $this->index) {
                    $clazz = $indexObj->getClass();
                    break;
                }
            }

            error_log('CLASS: ' . $clazz);

            $dataobject = DataObject::get_by_id($clazz, $assoc['id']);

            // Get highlight snippets, but only if a query parameter was passed in
            if (!empty($q)) {
                $snippets = Helper::create($connection)->callSnippets(
                // @todo get from index, need all text fields
                    $dataobject->Title . ' ' . $dataobject->Content,
                    //@todo hardwired
                    'sitetree_index',
                    $q,
                    [
                        'around' => 10,
                        'limit' => 200,
                        'before_match' => '<b>',
                        'after_match' => '</b>',
                        'chunk_separator' => '...',
                        'html_strip_mode' => 'strip',
                    ]
                )->execute()->getStored();
                $dataobject->Snippets = $snippets[0]['snippet'];
            }




            $formattedResult = new ArrayData([
                'Record' => $dataobject
            ]);

            $formattedResults->push($formattedResult);
        }

        $elapsed = round(microtime(true) * 1000) - $startMs;

        $pagination = new PaginatedList($formattedResults);
        $pagination->setCurrentPage($this->page);
        $pagination->setPageLength($this->pageSize);
        $pagination->setTotalItems($searchInfo['total_found']);



        return [
            'Records' => $formattedResults,
            'PageSize' => $this->pageSize,
            'Page' => $this->page,
            'TotalPages' => 1+round($searchInfo['total_found'] / $this->pageSize),
            'ResultsFound' => $searchInfo['total_found'],
            'Time' => $elapsed/1000.0,
            'Pagination' => $pagination,
            'AllFacets' => empty($facets) ? False : new ArrayList($facets)
        ];
    }
}


src/Service/Suggester.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 25/3/2561
 * Time: 1:35 น.
 */

namespace Suilven\SphinxSearch\Service;


use Foolz\SphinxQL\Drivers\Pdo\ResultSet;
use Foolz\SphinxQL\Facet;
use Foolz\SphinxQL\Helper;
use Foolz\SphinxQL\SphinxQL;
use SilverStripe\ORM\ArrayList;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\PaginatedList;
use SilverStripe\View\ArrayData;
use Suilven\FreeTextSearch\Indexes;

class Suggester
{

    private $client;

    private $index = 'sitetree';

    /**
     * @param string $index
     */
    public function setIndex($index)
    {
        $this->index = $index;
    }


    public function __construct()
    {
        $this->client = new Client();
    }


    public function suggest($q)
    {
        $suggestions = [];

        if (!empty($q)) {
            $connection = $this->client->getConnection();
            $e = $this->client->escapeSphinxQL($q);
            $indexName = $this->index . '_index';
            $query = SphinxQL::create($connection)->query("CALL QSUGGEST('$e', '{$indexName}')");
            $result = $query->execute()
                ->getStored();

           // @todo FIX Can we return multiple results and also can we pass in multiple words
            // result returns a string then a couple of numbers, no idea what the numbers are
            $suggestions = $result[0]['suggest'];

        }

        return [$suggestions];

    }
}


src/Service/Indexer.php
<?php
/**
 * Created by PhpStorm.
 * User: gordon
 * Date: 24/3/2561
 * Time: 21:14 น.
 */

namespace Suilven\SphinxSearch\Service;


use SilverStripe\Core\ClassInfo;
use SilverStripe\Core\Config\Config;
use SilverStripe\Core\Environment;
use SilverStripe\ORM\ArrayList;
use SilverStripe\ORM\DataList;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\DataObjectSchema;
use SilverStripe\ORM\DataQuery;
use SilverStripe\Versioned\Versioned;
use SilverStripe\View\ArrayData;
use Suilven\FreeTextSearch\Index;
use Suilven\FreeTextSearch\Indexes;

class Indexer
{
    protected $databaseName;

    protected $databaseHost;

    /**
     * @var null|Indexes indexes in current context
     */
    private $indexes = null;

    /**
     * Indexer constructor.
     * @param Indexes $indexes indexes in context
     */
    public function __construct($indexes)
    {
        $this->indexes = $indexes;
    }

    /**
     * @param mixed $databaseName
     */
    public function setDatabaseName($databaseName)
    {
        $this->databaseName = $databaseName;
    }

    /**
     * @param mixed $databaseHost
     */
    public function setDatabaseHost($databaseHost)
    {
        $this->databaseHost = $databaseHost;
    }

    /**
     * Generate config
     *
     * @return array of filename => sphinx config
     */
    public function generateConfig()
    {
        $allConfigs = [];


        /** @var Index $index */
        foreach($this->indexes as $index)
        {
            $className = $index->getClass();
            error_log("\n\n\n\n\n");

            error_log('>>>> CLASSNAME: ' . $className);

            error_log('INDEX: ' . print_r($index, 1));

            $name = $index->getName();
            $fields = []; // ['ID', 'CreatedAt', 'LastEdited'];


            // these are stored in the db but not part of free text search, a bit like tokens I guess
            $attributes = new ArrayList();

            // @todo different field types
            foreach($index->getFields() as $field)
            {
                $fields[] = $field;
            }

            foreach($index->getTokens() as $token)
            {
                $fields[] = $token;
            }

            $tokens = $index->getTokens();

            error_log('FIELDS AND TOKENS: ' . print_r($fields, 1));

            /** @var DataList $query */
            $singleton = singleton($className);
            $tableName = $singleton->config()->get('table_name');
            $schema = $singleton->getSchema();

            $specs = $schema->fieldSpecs($className, DataObjectSchema::DB_ONLY);


            // need to override sort, set it to null
            Config::modify()->set($className, 'default_sort', null);

            // @todo fix reference here
            /** @var DataObject $queryObject */

            /** @var $DataList $queryObject */
            $queryObject = Versioned::get_by_stage($className, Versioned::LIVE);


            error_log('CLASS:' . get_class($queryObject));


            error_log('QUERIED COLS FOR ' . $className . ': ' . print_r($fields, 1));

            // this is how to do it with a DataList, it clones and returns a new DataList
            $queryObject = $queryObject->setQueriedColumns($fields);


            // this needs massages for sphinx
            $sql = $queryObject->sql();

            error_log('SQL INIT: ' . $sql);


            $classNameInHierarchy = $className;


            $joinClasses = [];

            // need to know for the stage, dataobjects assumed flat
            $isSiteTree = false;
            while($classNameInHierarchy != 'SilverStripe\ORM\DataObject')
            {
                if ($classNameInHierarchy != 'SilverStripe\CMS\Model\SiteTree') {
                    $joinClasses[] = '\'' .  str_replace('\\', '\\\\', $classNameInHierarchy) . '\'';
                } else {
                    $isSiteTree = true;
                }

                error_log('CN:' . $classNameInHierarchy);
                $instance = new $classNameInHierarchy;
                $classNameInHierarchy = get_parent_class($classNameInHierarchy);
            }

            error_log('JOIN CLASSES: ' . print_r($joinClasses, 1));


            // error_log('SQL T1 ' . $sql);

            $sql = str_replace('"', '`', $sql);

            // need to move ID to first param
            if ($isSiteTree) {
                $sql = str_replace("`{$tableName}_Live`.`ID`, ", '', $sql);
                $sql = str_replace('SELECT DISTINCT', "SELECT DISTINCT `{$tableName}_Live`.`ID`, ", $sql);
            } else {
                $sql = str_replace("`{$tableName}`.`ID`, ", '', $sql);
                $sql = str_replace('SELECT DISTINCT', "SELECT DISTINCT `{$tableName}`.`ID`, ", $sql);
            }


            $commas = str_repeat('?, ', sizeof($joinClasses));
            $commas = substr( $commas, 0, -2 );
            $columns = implode(', ', $joinClasses);
            error_log('COMMAS: ' . $commas);
            $sql = str_replace('WHERE (`SiteTree_Live`.`ClassName` IN (' . $commas. '))',
                "WHERE (`SiteTree_Live`.`ClassName` IN ({$columns}))",
                $sql);
           // $sql = str_replace('WHERE (`SiteTree`.`ClassName` IN (?, ?))', "WHERE (`SiteTree`.`ClassName` IN ('{$joinClasses[0]}', '{$joinClasses[1]}'))", $sql);


            error_log('--------------------');
            error_log($sql);
            error_log($commas);
            error_log($columns);
            error_log('/--------------------');



            $sqlArray = explode(PHP_EOL, $sql);
            $sql = implode(' \\' . "\n", $sqlArray);

            // loop through fields adding attribute or altering sql as needbe
            $allFields = $fields;
            $allFields[] = 'LastEdited';
            $allFields[] = 'Created';

            // make modifications to query and or attributes but only if required
            foreach($allFields as $field)
            {
                if (isset($specs[$field])) {
                    $fieldType = $specs[$field];
                    error_log('FT:' . $field . ' --> ' . $fieldType);
                    switch($fieldType) {
                        case 'DBDatetime':
                            $sql = str_replace("`$tableName`.`$field`", "UNIX_TIMESTAMP(`$tableName`.`$field`) AS `$field`" , $sql);
                            // $sql = str_replace("`$tableName`.`$field`", "UNIX_TIMESTAMP(`$tableName`.`$field`) AS {$field}" , $sql);
                            $attributes->push(['Name' => $field, 'Type' => 'sql_attr_timestamp']);
                            break;
                        case 'Datetime':
                            $sql = str_replace("`$tableName`.`$field`", "UNIX_TIMESTAMP(`$tableName`.`$field`) AS `$field`" , $sql);
                            // this breaks order by if field is after: $sql = str_replace("`$tableName`.`$field`", "UNIX_TIMESTAMP(`$tableName`.`$field`) AS {$field}" , $sql);
                            $attributes->push(['Name' => $field, 'Type' => 'sql_attr_timestamp']);
                            break;
                        case 'Boolean':
                            $attributes->push(['Name' => $field, 'Type' => 'sql_attr_bool']); // @todo informed guess
                            break;
                        case 'ForeignKey':
                            $attributes->push(['Name' => $field, 'Type' => 'sql_attr_uint']);
                            break;
                        default:
                            // do nothing
                            break;
                    }

                    // strings and ints may need tokenized, others as above.  See http://sphinxsearch.com/wiki/doku.php?id=fields_and_attributes
                    if (in_array($field, $tokens)) {
                        $fieldType = $specs[$field];

                        // remove string length from varchar
                        if (substr( $fieldType, 0, 7 ) === "Varchar") {
                            $fieldTYpe = 'Varchar';
                        }

                        // @todo - float
                        // NOTE, cannot filter on string attributes, see http://sphinxsearch.com/wiki/doku.php?id=fields_and_attributes
                        // OH, it seems to work :)
                        switch($fieldType) {
                            case 'Int':
                                $attributes->push(['Name' => $field, 'Type' => 'sql_attr_uint']);
                                break;
                            case 'Varchar':
                                $attributes->push(['Name' => $field, 'Type' => 'sql_attr_string']);
                                break;
                            case 'HTMLText':
                                $attributes->push(['Name' => $field, 'Type' => 'sql_attr_uint']);
                                break;
                            case 'Float':
                                $attributes->push(['Name' => $field, 'Type' => 'sql_attr_float']);
                                break;
                            default:
                                // do nothing
                                break;
                        }
                    }
                } else {
                    user_error("The field {$field} does not exist for class {$className}");
                }

                //
            }

            /**
             * to add
             * 	sql_attr_string		= classname
             */


            $params = new ArrayData([
               'IndexName' => $name,
               'SQL' => 'SQL_QUERY_HERE',
                'DB_HOST' => !empty($this->databaseHost) ? $this->databaseHost : Environment::getEnv('SS_DATABASE_SERVER'),
                'DB_USER' => Environment::getEnv('SS_DATABASE_USERNAME'),
                'DB_PASSWD' => Environment::getEnv('SS_DATABASE_PASSWORD'),
                'DB_NAME' => !empty($this->databaseName) ? $this->databaseName : Environment::getEnv('SS_DATABASE_NAME'),
                'Attributes' => $attributes,
            ]);

            error_log('INDEXER PARAMS:' . print_r($params, 1));

            $configuraton = $params->renderWith('IndexClassConfig');



            $configuration2 = str_replace('SQL_QUERY_HERE', $sql, $configuraton);

            // @todo generic naming
            $allConfigs[$name] = "{$configuration2}";
        }
        return $allConfigs;
    }

    /**
     * Create a valid sphinx.conf file and save it.  Note that the commandline or web server user must have write
     * access to the path defined in _config.
     */
    public function saveConfig()
    {
        // specific to the runnnig of sphinx
        $common = file_get_contents( __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . '..'
            . DIRECTORY_SEPARATOR . 'sphinxconfig' . DIRECTORY_SEPARATOR . 'common.conf');
        $indexer = file_get_contents( __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . '..'
            . DIRECTORY_SEPARATOR . 'sphinxconfig' . DIRECTORY_SEPARATOR . 'indexer.conf');
        $searchd = file_get_contents( __DIR__ . DIRECTORY_SEPARATOR . '..' . DIRECTORY_SEPARATOR . '..'
            . DIRECTORY_SEPARATOR . 'sphinxconfig' . DIRECTORY_SEPARATOR . 'searchd.conf');


        // specific to silverstripe data
        $sphinxConfigurations = $this->generateConfig();
        $sphinxSavePath = Config::inst()->get('Suilven\SphinxSearch\Service\Client', 'config_file');

        $config = $common . $indexer . $searchd;

        foreach(array_keys($sphinxConfigurations) as $filename) {
            $config .= $sphinxConfigurations[$filename];
        }

        file_put_contents($sphinxSavePath, $config);

        error_log('CONFIG SAVED: ' . $config);
    }
}
